concordium-node/src/grpc2.rs

```diff
@@ -985,6 +988,57 @@ pub mod server {
             let response = tonic::Response::new(receiver);
             Ok(response)
         }
+
+        async fn peer_connect(
+            &self,
+            request: tonic::Request<crate::grpc2::types::PeerConnection>,
+        ) -> Result<tonic::Response<crate::grpc2::types::BooleanResponse>, tonic::Status> {
+            if self.node.is_network_stopped() {
+                Err(tonic::Status::failed_precondition(
+                    "The network is stopped due to unrecognized protocol update.",
+                ))
+            } else {
+                let peer_connect = request.into_inner();
+                if let Ok(ip) = <std::net::IpAddr as std::str::FromStr>::from_str(
+                    &peer_connect.ip.require()?.value,
+                ) {
+                    let addr = SocketAddr::new(ip, peer_connect.port.require()?.value as u16);
+                    self.node.register_conn_change(crate::connection::ConnChange::NewConn {
+                        addr,
+                        peer_type: crate::common::PeerType::Node,
+                        given: true,
+                    });
+                    Ok(tonic::Response::new(crate::grpc2::types::BooleanResponse {
+                        value: true,
+                    }))
+                } else {
+                    Err(tonic::Status::invalid_argument("Invalid IP address"))
+                }
+            }
+        }
+
+        async fn peer_disconnect(
+            &self,
+            request: tonic::Request<crate::grpc2::types::PeerConnection>,
+        ) -> Result<tonic::Response<crate::grpc2::types::BooleanResponse>, tonic::Status> {
+            if self.node.is_network_stopped() {
+                Err(tonic::Status::failed_precondition(
+                    "The network is stopped due to unrecognized protocol update.",
+                ))
+            } else {
+                let peer_connect = request.into_inner();
+                if let Ok(ip) = <std::net::IpAddr as std::str::FromStr>::from_str(
+                    &peer_connect.ip.require()?.value,
```

Comment:
Recommend that you add a `TryFrom<PeerConnection> for SocketAddr`. Then this becomes a lot simpler.